### DISCUSSION TITLE
Embedding with different roles

### DISCUSSION THREAD

**Discussion by IIppolitov** [apache/superset#28697] [Ideas]
It would be great to have the ability to embed dashboards using different Roles for the guest token.
During the embedment we generate a guest token, which takes the Role from the GUEST_ROLE_NAME config.
But sometimes it would be great to have the ability to clearly indicate which role we would like to use for the embedment.
The main reasoning behind it is that depending on the dashboard that is embedded we would like to give the ability for users to download (or to forbid the download) excel and csv functionality (due to local Personal Information processing laws excel download from inside the system is a security breach, which we legally are obliged to eliminate).

The best way to do it, in my opinion, is to expand the GuestTokenUser class with a new attribute "Role" and check in get_guest_user_from_token if the role attribute is present and find the according role (if the attribute is missing - use the configuration one).

**Comment by arafoperata** [comment]
Would also like to see this. Should be a straightforward implementation, and makes sense that we might want guest users with different permissions. Happy to contribute a PR.

**Reply by dmalav** [reply]
We are looking for exactly same functionality, let's contribute and make it happen.

**Comment by Nandhan007** [comment]
Can you help me when i am trying to get access token using login api call but it actually works for admin role and i will get access token response but for other roles, i am actually getting 401 unauthorised error. I need a solution for this problem

**Comment by mschaefer-gresham** [comment]
I just asked the same [question](https://github.com/apache/superset/discussions/30420).

Would love to see this.

**Reply by Nandhan007** [reply]
I found a solution actually If i try to get access token with superset inbuilt API uri It will respond to the local users only. For the local users what I mean is users registered via superset without using third party registration for authentication and authorization. Actually I am facing issue when I try to integrate with keycloak third party IAM tool but while embed the dashboard it will allow the users who registered via superset not a keycloak IAM tool. Finally problem is resolved.

**Comment by jpchev** [comment]
I have the same need, this functionality would be extremely useful to implement a powerful embedding

**Reply by teharris1** [reply]
I posted a potential solution here:
https://github.com/apache/superset/discussions/30420
